internal/gatewayapi/translator.go

```diff
@@ -777,6 +899,71 @@ func isValidCrossNamespaceRef(from crossNamespaceFrom, to crossNamespaceTo, refe
 	return false
 }
 
+// Checks if a hostname is valid according to RFC 1123 and gateway API's requirement that it not be an IP address
+func isValidHostname(hostname string) (valid bool, reason string) {
+	if len(hostname) == 0 || hostname == "" {
+		return false, "Hostname cannot be empty"
+	}
+
+	if len(hostname) > 255 {
+		return false, fmt.Sprintf("Hostname '%s' has a length of %d. cannot exceed 255", hostname, len(hostname))
+	}
+
+	if hostname[0] == '-' || hostname[0] == '.' {
+		return false, "Hostname cannot start with '.' or '-'"
+	}
+	if hostname[len(hostname)-1] == '-' || hostname[len(hostname)-1] == '.' {
+		return false, "Hostname cannot end with '.' or '-'"
+	}
+
+	// The only allowed characters in hostnames are a-z, A-Z, 0-9, ., and -
+	var validChar = regexp.MustCompile(`[a-zA-Z0-9.-]`).MatchString
+
+	curLabel := 0
+	for i := 0; i < len(hostname); i++ {
```

Comment:
Any reason not to use `for i, char := range hostname {`?

Reply:
Probably just had a brainfart thinking about indexes too much